design/src/rfcs/rfc0004_retry_behavior.md

```diff
@@ -0,0 +1,147 @@
+use aws_types::config::Config;
+use aws_types::retry_config::RetryConfig;
+
+async fn create_config() -> aws_types::config::Config {
+    let retry_config = RetryConfig::builder().max_attempts(5).build();
+    let config = aws_config::from_env()
+        .retry_config(retry_config)
+        .load().await;
+
+    config
+}
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = create_config().await;
+
+    let sts = sts::Client::new(&config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+## Calling a method on service-specific config
+
+Here's an example app that creates a service-specific config with custom retry behavior and then logs your AWS user's identity
+
+```rust
+use aws_sdk_sts as sts;
+use aws_types::config::Config;
+use aws_types::retry_config::RetryConfig;
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = aws_config::load_from_env().await;
+    let retry_config = RetryConfig::builder().max_attempts(5).build();
+    let sts_config = sts::config::Config::from(&config).retry_config(retry_config).build();
+
+    let sts = sts::Client::new(&sts_config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+## Disabling retries
+
+Here's an example app that creates a service-specific config with custom retry behavior disabling retries and then logs your AWS user's identity
+
+```rust
+use aws_sdk_sts as sts;
+use aws_types::config::Config;
+use aws_types::retry_config::RetryConfig;
+
+#[tokio::main]
+async fn main() -> Result<(), sts::Error> {
+    let config = aws_config::load_from_env().await;
+    let retry_config = RetryConfig::builder().max_attempts(0).build();
+    let sts_config = sts::config::Config::from(&config).retry_config(retry_config).build();
+
+    let sts = sts::Client::new(&sts_config);
+    let resp = sts.get_caller_identity().send().await?;
+    println!("your user id: {}", resp.user_id.unwrap_or_default());
+    Ok(())
+}
+```
+
+Behind the scenes
+-----------------
+
+// TODO
+
+Changes Checklist
+-----------------
+
+- [ ] Create new Kotlin decorator `RetryDecorator`
+  - Based on [RegionDecorator.kt](https://github.com/awslabs/smithy-rs/blob/main/aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/RegionDecorator.kt)
+- [ ] Create `aws_types::retry_config::RetryConfig` struct and corresponding builder with a `max_attempts` setter.
+- [ ] Create `aws_config::meta::retry_config::RetryConfigProviderChain`
+- [ ] Create `aws_config::meta::retry_config::ProvideRetryConfig`
+- [ ] Create `EnvironmentVariableRegionProvider` struct
+- [ ] Add `retry_config` method to `aws_config::ConfigLoader`
+- [ ] Update `AwsFluentClientDecorator` to correctly configure the retry behavior of its inner `aws_hyper::Client` based on the retry config.
+- [ ] Add tests
```

Comment:
What kind of tests do we have for retries currently? Will these be easy to augment for this?

Reply:
I believe that all the retry-related tests are here: https://github.com/awslabs/smithy-rs/blob/05d9ec9afec82880204bfff328f10ccf444b1df0/rust-runtime/smithy-client/src/retry.rs#L309

I can see adding other tests that mirror ones in 
- `EnvironmentVariableRegionProvider`
- aws/rust-runtime/aws-config/src/meta/region.rs
- probably more places that I've forgotten right now

All the test will fall under 2 categories:
1. tests that test the loading/storage/passing of configuration
2. tests that test retrying HTTP calls

I'm hoping that for type 1 tests, I can copy and modify tests we already have.
I think that for type 2 tests I can do some copying but will have to write some new ones as well.